040-SDK/150-file-attachments.mdx

```diff
@@ -7,6 +7,91 @@ slug: sdk/file-attachments
+
+- [Upload URLs](#upload-urls), when requested from an existing file object, provide a temporary, secure URL to upload files to. They can be used client-side and are good when dealing with large files through a client interface like a browser. This method comes at the downside of requiring extra steps across the client and server.
+- [Record APIs](#record-apis) provide the easiest way to upload files, and are the only methods that allow uploading files during the same step as creating a record. The downside is that they only support small files and will likely run over the limits of serverless functions from providers like Vercel.
+- [Binary APIs](#file-binary-apis) allow for uploading large files on an existing record. Unfortunately, unlike Upload URLs, they must be called server side, which will likely run over the limits of serverless functions from providers like Vercel. This makes their usefulness limited to scenarios where you control the environment completely, like uploading large files from a local machine.
+
+Xata recommends using Upload URLs for most scenarios. Record and Binary APIs exist for more specific situations where you know the size of your data or have complete control over the hosting infrastructure.
+
+## Upload URLs
+
+Xata provides a secure method to gain temporary upload URLs to upload or update files on existing records. They are the recommended pattern for uploading large files through a web browser or when you want to transfer a file client side. Upload URLs allow you to bypass the size restrictions of many server-side function hosts like Vercel and send files directly to Xata within its larger [REST API limits](/docs/rest-api/limits#file-attachments-limits).
+
+The only negative to using upload URLs versus the file or binary APIs is that this method requires multiple steps (creating, then updating the record), usually across both the server and the client.
+
+A typical pattern for a JavaScript application follows:
+
+1. On the server side create a record with an empty file using `xata.db.myTableName.create()`, simultaneously requesting an `uploadUrl` from that newly created, but empty file.
+2. On the server, pass a response with the file's `uploadUrl` to the client.
+3. On the client, upload the file using a `PUT` request against the `uploadUrl` given, making sure to account for any cleanup necessary if the upload fails.
+
+A server-side API route (heavily simplified, and not accounting for your particular framework) might look like this:
+
+```ts title="app/api/create-record.ts"
+// ... Your api route code
+
+// Create an empty record with no base64 content on a `myTableName` table. The column for the file is `myFileColumnName`.
+const record = await xata.db.myTableName.create(
+  { name, myFileColumnName: { name: name, mediaType: fileType, base64Content: '' } },
+  // Request an uploadUrl from the created record. We'll use this client-side to update the record.
+  ['*', 'myFileColumnName.uploadUrl']
```

Comment:
Do we want to encourage the use of `*` + generated fields as a pattern for retrieving file objects?
Otherwise the start `*` is not needed here as we only need the upload url in order to upload in the client function below.

Reply:
That's true. I'll update.